File: neetcode150/stack/leetcode155.py
class MinStack:
    def __init__(self):
        self.stack = []

    def push(self, val: int) -> None:
        if not self.stack:
            self.stack.append((val, val))
            return

        _min = self.getMin()
        self.stack.append((val, min(_min, val)))

    def pop(self) -> None:
        self.stack.pop()

    def top(self) -> int:
        top, _ = self.stack[-1]
        return top

    def getMin(self) -> int:
        _, _min = self.stack[-1]
        return _min

File: js_algorithms/neetcode150/stack/leetcode155.js
const MinStack = function () {
  this.stack = [];
};

/**
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function (val) {
  if (this.stack.length) {
    const min = this.getMin();
    this.stack.push([val, Math.min(min, val)]);
  } else {
    this.stack.push([val, val]);
  }
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function () {
  this.stack.pop();
};

/**
 * @return {number}
 */
MinStack.prototype.top = function () {
  const last = this.stack[this.stack.length - 1];
  return last[0];
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function () {
  const last = this.stack[this.stack.length - 1];
  return last[1];
};
